### ISSUE TITLE
508-defect-2 [COGNITION, SCREENREADER]: Pagination SHOULD use meaningful landmarks and aria techniques

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#17085]
# [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

<hr/>

## Point of Contact

**VFS Point of Contact:**  Josh

## User Story or Problem Statement

As a screen reader user, I want to know when I am using a pagination and what page I am currently on so that I can navigate the payment tables.

## Details

Several issues currently exist with the pagination including:
- It is not discoverable by landmark or navigation shortcuts as it is not wrapped in a `nav`.
- If two exist on the page, screen reader users may confuse them with one another as they are not labelled and described appropriately.
- No semantic equivalent to the visual affordance of a current page exists.
- Screen reader users won't know how many items exist in the pagination as pagination links are not written as an unordered list.

I've documented a remediation below. **However, this may be worth re-exploring from a design lens in the future.** A pagination with listed numbers may not be the most optimal approach:
- Could we simply use next and previous? 
- Do users have a reason to jump by numbers on a page?
  - If yes, are there better methods of helping them navigate to what they're looking for? e.g. search, filters

## Tasks
- [ ] Obtain guidance from platform (tag Ryan Thurwell/Chris Valerida/Brook Johnson/ C Wong) on what should be done with this ticket on Josh's behalf

## Acceptance Criteria
- [ ] A path forward has been established

**Next steps**
- [ ] Paginations are wrapped in a `nav` element.
- [ ] Paginations are labeled as paginations using `aria-label`.
- [ ] Paginations are described by their respective table headings using `aria-describedby`.
- [ ] Current page links are noted using `aria-current`.
- [ ] Pagination is written as an unordered list instead of a series of `div` and `a` elements.

## Environment

* Operating System: Mac, Windows
* Browser: Safari, Firefox
* Screenreading device: VoiceOver, NVDA
* Server destination: staging, production

## Steps to Recreate

1. Log in as user 54.
1. Enter `https://staging.va.gov/va-payment-history/payments/` in browser.
1. Start up the screenreader of your choice.
1. Tab to the pagination.
1. NVDA announces the first link as "Load page 1 link".
1. VoiceOver announces "Leaving...." "Load page 1 group".

### Current
``` html
<div class="va-pagination vads-u-border-top--0">
  <span class="va-pagination-prev"></span>
  <div class="va-pagination-inner">
    <a class="va-pagination-active" aria-label="Load page 1 " tabindex="0">1</a>
    <a class="" aria-label="Load page 2 " tabindex="0">2</a>
   <!-- anchor elements continue -->
    <span>
      <a aria-label="...">...</a>
      <a aria-label="Load last page ">12</a>
    </span>
  </div>
  <span class="va-pagination-next">
    <a aria-label="Load next page " tabindex="0">Next</a>
  </span>
</div>
```

### 1: Wrap the pagination in a descriptive `nav`
- Wrap the pagination in a `nav` to make it an accessible landmark.
- Add `aria-label="pagination"` so it is announced as a pagination to the user.
- Add `aria-describedby=""` so that the user knows which pagination they are using.

``` diff
+ <nav aria-label="pagination" aria-describedby="paymentsRecievedHeader OR paymentsReturnedHeader">
  <div class="va-pagination vads-u-border-top--0">
    <span class="va-pagination-prev"></span>
    <div class="va-pagination-inner">
      <a class="va-pagination-active" aria-label="Load page 1 " tabindex="0">1</a>
      <a class="" aria-label="Load page 2 " tabindex="0">2</a>
     <!-- anchor elements continue -->
      <span>
        <a aria-label="...">...</a>
        <a aria-label="Load last page ">12</a>
      </span>
    </div>
    <span class="va-pagination-next">
      <a aria-label="Load next page " tabindex="0">Next</a>
    </span>
  </div>
+ </nav>
```

### 2: Convert the pagination into an unordered list
This will allow the user to know how many pages exist to navigate through.

``` diff
<nav aria-label="pagination" aria-describedby="paymentsRecievedHeader OR paymentsReturnedHeader">
  <div class="va-pagination vads-u-border-top--0">
+ <ul>
     <li><span class="va-pagination-prev"></span></li>
-     <div class="va-pagination-inner">
+   <li><a class="va-pagination-active" aria-label="Load page 1 " tabindex="0">1</a></li>
+   <li><a class="" aria-label="Load page 2 " tabindex="0">2</a></li>
     <!-- anchor elements continue -->
-      <span>
+   <li><a aria-label="...">...</a></li>
+   <li><a aria-label="Load last page ">12</a></li>
-      </span>
-    </div>
-    <span class="va-pagination-next">
+   <li><a aria-label="Load next page " tabindex="0">Next</a></li>
-    </span>
+ </ul>
  </div>
</nav>
```

### 3: Add `aria-current` to the current page link
This will announce to the user if they are focused on a link that contains the current page.
``` diff
<nav aria-label="pagination" aria-describedby="paymentsRecievedHeader OR paymentsReturnedHeader">
  <div class="va-pagination vads-u-border-top--0">
    <ul>
      <li><span class="va-pagination-prev"></span></li>
+     <li><a class="va-pagination-active" aria-label="Load page 1 " tabindex="0" aria-current="page">1</a></li>
      <li><a class="" aria-label="Load page 2 " tabindex="0">2</a></li>
      <!-- anchor elements continue -->
      <li><a aria-label="...">...</a></li>
      <li><a aria-label="Load last page ">12</a></li>
      <li><a aria-label="Load next page " tabindex="0">Next</a></li>
    </ul>
  </div>
</nav>
```

## Outstanding Questions @1Copenut 
- Do we have documentation on the history of the pagination? [The existing example on the design system is not accessibly written.](https://design.va.gov/components/pagination)
- What was the purpose behind adding `tab-index` for the pagination links?
- Why was the pagination not written as an unordered list? Are there constraints I'm unaware of that is barring the recommended solution? 
- There is also a focus management issue where upon selecting "Next" on the page before the final page, focus is lost as "Next" is removed.

## WCAG or Vendor Guidance

* [Using the aria current attribute - Leonie Watson](https://tink.uk/using-the-aria-current-attribute/)
* [A11y style guide - pagination](https://a11y-style-guide.com/style-guide/section-navigation.html#kssref-navigation-pagination)

## Screenshots or Trace Logs

<img width="1076" alt="Screen Shot 2020-12-08 at 9 10 40 AM" src="https://user-images.githubusercontent.com/14154792/101503584-3f711d80-3940-11eb-8974-5ec72a1647a5.png">

**Comment by jason-gcio** [comment]
We will present this to platform and get some next steps from them.

**Comment by cohnjesse** [comment]
@joshkimux This is a design system component and should be handled by the VSP team.

**Comment by joshkimux** [comment]
@cohnjesse Agreed! Would y'all be ok with leaving this ticket open until VSP is able to address it? Just so we don't lose track of it

**Comment by jason-gcio** [comment]
Tracked via platform: https://dsva.slack.com/archives/CBU0KDSB1/p1611867534443500
